### PR TITLE
Converter for TRT NMS using TRT plugin registry

### PR DESCRIPTION
**Pull request by pooyadavoodi**
- Load TRT plugin registry dynamically (dlopen).
- Initialize TRT plugin registry always (even if the model doesn't need any plugin).
- Add a converter for CombinedNonMaxSuppression
- Add unit tests.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc [hunk 1]
```diff
@@ -957,6 +961,26 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
   LOG(INFO) << "Number of TensorRT candidate segments: "
             << initial_segments.size();
 
+  // Check if plugins can be aaccessed.
+  int num_trt_plugins = 0;
+  nvinfer1::IPluginCreator* const* trt_plugin_creator_list =
```
Comment (pooyadavoodi) [inline tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: Need to run some tests to ensure this works with TRT4.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 2]
```diff
@@ -2229,6 +2229,114 @@ TEST_F(OpConverterTest, ConvertSquare) {
...
+//
+//      auto nms_op = ops::CombinedNonMaxSuppression(s.WithOpName("my_nms"),
+//                                                   boxes_tensor,
+//                                                   scores_tensor,
+//                                                   max_output_size_per_class,
+//                                                   max_total_size,
+//                                                   iou_threshold,
+//                                                   score_threshold,
+//                                                   nms_attrs);
+//      return nms_op.operation.node()->def();
```
Comment (pooyadavoodi) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: I am still working on this.

This is the line I get a crash. `node()` returns null pointer.

File: tensorflow/python/compiler/tensorrt/trt_convert.py [hunk 3]
```diff
@@ -482,6 +483,18 @@ def __init__(self,
...
           [str(x) for x in loaded_version]))
 
+    # Load TensorRT plugins.
+    try:
+      plugin_lib = ctypes.CDLL("libnvinfer_plugin.so")
+    except Exception as e:
+      tf_logging.warn("Failed to load libnvinfer_plugin.so" +  str(e))
+    else:
+      # Initialize and register TensorRT plugins.
+      plugin_lib_registered = plugin_lib.initLibNvInferPlugins(None, "")
```
Comment (pooyadavoodi) [inline tensorflow/python/compiler/tensorrt/trt_convert.py]: Need to run some tests to ensure this works with TRT4.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 4]
```diff
@@ -3634,10 +3634,163 @@ tensorflow::Status ConvertTopK(OpConverterParams* params) {
   return tensorflow::Status::OK();
 }
 
+#if NV_TENSORRT_MAJOR > 5 || (NV_TENSORRT_MAJOR == 5 && NV_TENSORRT_MINOR >= 1)
+tensorflow::Status ConvertCombinedNMS(OpConverterParams* params) {
+  const auto& inputs = params->inputs;
+  const auto& node_def = params->node_def;
+
+  if (inputs.size() != 6) {
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Please use `CheckInputsWeights()` in line 1324 to replace line 3642-3656.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 5]
```diff
@@ -2229,6 +2229,114 @@ TEST_F(OpConverterTest, ConvertSquare) {
...
+//  };
+//
+//  struct TestParams {
+//    TestParams(const std::vector<int32>& boxes_tensor_dims,
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: We don't need the constructor, aggregate initialization is available by default.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 2 above)
Comment (aaroey) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: This is super weird. I'll try probably next week to see what happens.

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc [hunk 6]
```diff
@@ -957,6 +961,26 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
...
+    getPluginRegistry()->getPluginCreatorList(&num_trt_plugins);
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: nit: please fix the indention and other formatting issues like line to long in the PR.

**Review by aaroey — CHANGES_REQUESTED** [review]

**Comment by FamousDirector** [comment]
@pooyadavoodi Would this speed up the `NonMaxSuppressionV3` layer (implemented via `tf.image.non_max_suppression`)? My application is currently bottlenecked by this layer.

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc (hunk 1 above)
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: This ConvertAfterShapes() is too large, can we encapsulate the newly added lines into a helper method and call it here?

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc [hunk 7]
```diff
@@ -96,6 +97,9 @@ Status TrtCandidateSelector::IsTensorRTCandidate(const tensorflow::Node* node) {
       "AvgPool",
       "BatchMatMul",
       "BiasAdd",
+#if NV_TENSORRT_MAJOR > 5 || (NV_TENSORRT_MAJOR == 5 && NV_TENSORRT_MINOR >= 1)
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc]: Please help to resolve the conflicts.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 2 above)
Comment (aaroey) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: Hi Pooya, I copied this portion of the code (without the actual tests below) but cannot reproduce the problem. Would you please uncomment all of them, and fix the formats? Then once RC1 is out I can patch and test again.

Thanks.

File: tensorflow/compiler/tf2tensorrt/segment/segment.cc [hunk 8]
```diff
@@ -450,6 +450,12 @@ tensorflow::Status SegmentGraph(
         num_unsupported_ops++;
         node = nullptr;
       }
+      else {
+        VLOG(2) << "Accepted as a TF-TRT candidate, "
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/segment/segment.cc]: Why do we need this log?

File: tensorflow/compiler/tf2tensorrt/segment/segment.cc [hunk 9]
```diff
@@ -450,6 +450,12 @@ tensorflow::Status SegmentGraph(
...
+                << "(Op type: " << node->tf_node()->type_string() << "), "
+                << "(Op name: " << node->name() << "), "
+                << "(Reason: " << status << ")";
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/segment/segment.cc]: status will always be ok I guess.

File: tensorflow/python/compiler/tensorrt/trt_convert.py [hunk 10]
```diff
@@ -482,6 +483,18 @@ def __init__(self,
       tf_logging.info("Running against TensorRT version %s" % ".".join(
           [str(x) for x in loaded_version]))
 
+    # Load TensorRT plugins.
+    try:
+      plugin_lib = ctypes.CDLL("libnvinfer_plugin.so")
```
Comment (aaroey) [inline tensorflow/python/compiler/tensorrt/trt_convert.py]: Why do we need to load it here? Can we just link the library into convert_nodes?

**Review by aaroey — CHANGES_REQUESTED** [review]
Thanks a lot for the PR!

**Review by aaroey — DISMISSED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 11]
```diff
@@ -3914,9 +3950,183 @@ Status ConvertTopK(OpConverterParams* params) {
...
+  if (pad_per_class) {
+    topK = std::min(max_size_per_class * num_classes, max_total_size);
+  } else {
+    topK = max_total_size;
+  }
+  const int keepTopK = topK;
+  float score_thresh =
+      *(static_cast<float*>(const_cast<void*>(score_threshold.GetValues())));
+  const int background_id = -1;
+  nvinfer1::PluginField fields[7] = {
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: @pooyadavoodi Hi Pooya, would you please help to add the `is_normalized` attribute/field you mentioned to fix the accuracy issue? Thanks.

**Review by aaroey — DISMISSED** [review]

**Comment by rthadur** [comment]
@pooyadavoodi can you please check failed build

**Review by aaroey — APPROVED** [review]

**Comment by aaroey** [comment]
@rthadur I'll handle that.

**Pull request merged**